ext/drx/drx.c

```diff
@@ -2429,3 +2441,363 @@ drx_expand_scatter_gather(void *drcontext, instrlist_t *bb, OUT bool *expanded)
+                    if (opnd_is_reg(src0)) {
+                        reg_id_t tmp_gpr = opnd_get_reg(src0);
+                        if (reg_is_gpr(tmp_gpr)) {
+                            if (restore_scratch_mask_start <= info->raw_mcontext->pc &&
+                                info->raw_mcontext->pc <= prev_pc) {
+                                /* The scratch mask is always k0. this is hard-coded
+                                 * in drx. We carefully only update the lowest 16 bits
+                                 * because the mask was saved with kmovw.
+                                 */
+                                ASSERT(sizeof(info->mcontext->opmask[0]) ==
+                                           sizeof(long long),
+                                       "internal error: unexpected opmask slot size");
+                                info->mcontext->opmask[0] &= ~0xffffLL;
+                                info->mcontext->opmask[0] |=
+                                    reg_get_value(gpr_save_scratch_mask, info->mcontext) &
+                                    0xffff;
+                                advance_state(
+                                    &detect_state,
+                                    DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
+                                    &allow_for_unknown_instr_count);
+                            }
+                        }
+                    }
+                }
+            }
+            if (allow_for_unknown_instr_inc(&allow_for_unknown_instr_count))
+                break;
+            advance_state(&detect_state, DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
+                          &allow_for_unknown_instr_count);
+            break;
+        default: ASSERT(false, "internal error: invalid state.");
+        }
+    }
+}
+
+static bool
+drx_event_restore_state(void *drcontext, bool restore_memory,
+                        dr_restore_state_info_t *info)
+{
+    instr_t inst;
+    if (info->fragment_info.cache_start_pc == NULL)
+        return true; /* fault not in cache */
+    instr_init(drcontext, &inst);
+    byte *pc = decode(drcontext, dr_fragment_app_pc(info->fragment_info.tag), &inst);
+    if (pc != NULL) {
+        if (instr_is_gather(&inst)) {
+            /* XXX i#2985: if !info->fragment_info.app_code_consistent, we run the state
+             * machine anyway.
+             */
+            scatter_gather_info_t sg_info;
+            get_scatter_gather_info(&inst, &sg_info);
+            if (sg_info.is_evex) {
+                drx_try_to_detect_avx512_gather_sequence(drcontext, info, &inst,
+                                                         &sg_info);
+            } else {
+                /* TODO i#2985: support AVX2 gather. */
+            }
+        } else if (instr_is_scatter(&inst)) {
+            /* TODO i#2985: support AVX-512 scatter. */
+        }
+    }
+    instr_free(drcontext, &inst);
+    return true;
```

Comment:
Doesn't it seem safer to return false on the failures cases in drx_try_to_detect_avx512_gather_sequence?  For restore_memory=false anyway.

Reply:
We are never restoring memory. Why do you think we should abort if restore_memory=false? Just making sure I understand.

So drx_try_to_detect_avx512_gather_sequence may fail to detect, but if we return false, wouldn't we fail every translation that is not a gather sequence? We could never translate?

Reply:
> We are never restoring memory. Why do you think we should abort if restore_memory=false? Just making sure I understand.

This callback should never abort when restore_memory=true is the point of checking restore_memory.  It should only consider refusing to translate when *not* restoring memory.

Reply:
Ok, that makes more sense, I had misread.

Reply:
> So drx_try_to_detect_avx512_gather_sequence may fail to detect, but if we return false, wouldn't we fail every translation that is not a gather sequence? We could never translate?

Yes, right.  What some clients do is to track all bbs of a certain type with a hashtable (handling replacement for new bbs).  Then you would know for sure it's yours, and then you could treat errors as errors.

Reply:
Is this worth implementing here, did you want that? I didn't think this is neessary, because we decode the _app_ instruction. If this was a gather/scatter instruction, only then is it worth to start searching for the sequence. Otherwise we don't. Another easy improvement would also be to set a global flag once somebody used the expand_scatter_gather function. Because if nobody had called it, we don't need to bother to search. WDYT?

Reply:
Not sure, maybe not.  Decoding the app instr is not always possible: not just self-modifying code but IIRC fragments that have been flushed (not yet freed of course) which can be for a variety of reasons, including adaptive instrumentation (i.e., client just wants to replace instru).  (Though it seems like it should be "flushed and exec area removed" which won't happen on client flush...but glancing at the code it only checks FRAG_WAS_DELETED so it can't tell the difference.)

Reply:
I already have a flag expand_scatter_gather_drreg_initialized. That was an easy improvement. 

I think it is relatively efficient now as-is. Sure it could happen that somebody is using expand_scatter_gather() only on some instructions, yet we're now always searching the fragment as soon as the original app instruction was a scatter/gather even though it may not have been expanded.